### DISCUSSION TITLE
How would you recommend integrating with StreamJsonRpc ?

### DISCUSSION THREAD

**Discussion by liiri** [Tyrrrz/CliWrap#136] [Q&A]
My dotnet program is starting a subprocess and need to communicate back and forth with it using JSON-RPC.

I'm using [vs-streamjsonrpc](https://github.com/microsoft/vs-streamjsonrpc), the interface for creating a new rpc connection is accepting two streams - one to read from and one to write to: `Attach(Stream? sendingStream, Stream? receivingStream)`.

With plain `Process` my code looks something like this:

```cs
var process = new Process();
process.StartInfo = new ProcessStartInfo(...);
process.StartInfo.RedirectStandardInput = true;
process.StartInfo.RedirectStandardOutput = true;
process.Start();

var rpc = JsonRpc.Attach(sendingStream: process.StandardInput.BaseStream, receivingStream: process.StandardOutput.BaseStream);
```

How would you suggest to use `CliWrap` in this scenario?

I saw `PipeSource.FromStream()` and `PipeTarget.ToStream()`, but I think they're expect to read/write entire stream and not act for real time message passing between dotnet and the subprocess.

**Comment by Tyrrrz** [comment]
I don't know what to recommend not knowing the API you're working with (I haven't worked with `vs-streamjsonrpc`). At least with the `JsonRpc.Attach(...)` method I don't see how to do it. But in any case, I'd probably look into the direction of creating your own implementation of `PipeTarget` and `PipeSource`, those abstractions deal directly with process' streams.
